storage/range.go

```diff
@@ -1629,28 +1629,84 @@ func (r *Range) FirstIndex() (uint64, error) {
-	// TODO(bdarnell): package up all the data in the range, not just
-	// the descriptor (and send the appropriate log index).
-	data, err := gogoproto.Marshal(r.Desc())
+	// Copy all the data from a consistent RocksDB snapshot into a RaftSnapshotData.
+	snap := r.rm.NewSnapshot()
+	defer snap.Stop()
+	var snapData proto.RaftSnapshotData
+
+	// Certain keys are treated specially while building the snapshot;
+	// see the switch statement below for details.
+	hardStateKey := engine.RaftHardStateKey(r.Desc().RaftID)
+	appliedKey := engine.RaftAppliedIndexKey(r.Desc().RaftID)
+	descKey := engine.RangeDescriptorKey(r.Desc().StartKey)
+
+	// Save the range metadata while iterating. Use the values read from
+	// the snapshot instead of the members of the Range struct because
+	// they might be changed concurrently.
+	appliedIndex := uint64(raftInitialLogIndex)
+	var desc proto.RangeDescriptor
+	seenDesc := false
+
+	// Iterate over all the data in the range, including local-only data like
+	// the response cache.
+	for iter := newRangeDataIterator(r, snap); iter.Valid(); iter.Next() {
+		key, _, isValue := engine.MVCCDecodeKey(iter.Key())
+		switch {
+		case bytes.Equal(key, hardStateKey):
+			// The raft HardState must not be transmitted via snapshot,
+			// because the receiving node may have already cast a vote in this term.
+			continue
+		case bytes.Equal(key, appliedKey):
+			// Raft uses the applied index to synchronize the snapshot with the log.
+			var v proto.MVCCMetadata
+			err := gogoproto.Unmarshal(iter.Value(), &v)
+			if err != nil {
+				return raftpb.Snapshot{}, err
+			}
+			_, appliedIndex = encoding.DecodeUint64(v.Value.Bytes)
+		case bytes.Equal(key, descKey) && isValue && !seenDesc:
+			// Extract the descriptor so we can tell raft about the range membership at
+			// the time of the snapshot.
+			// TODO(bdarnell): Correctly handle pending transactions on the descriptor.
+			// Should we just skip the descriptor while iterating and call engine.MVCCGet
+			// on the snapshot?
+			seenDesc = true
+			var v proto.MVCCValue
+			err := gogoproto.Unmarshal(iter.Value(), &v)
+			if err != nil {
+				return raftpb.Snapshot{}, err
+			}
+			err = gogoproto.Unmarshal(v.Value.Bytes, &desc)
+			if err != nil {
+				return raftpb.Snapshot{}, err
+			}
+		}
+
+		snapData.KV = append(snapData.KV,
+			&proto.RaftSnapshotData_KeyValue{Key: iter.Key(), Value: iter.Value()})
+	}
+
+	data, err := gogoproto.Marshal(&snapData)
 	if err != nil {
 		return raftpb.Snapshot{}, err
 	}
 
-	// Synthesize our raftpb.ConfState from Desc.
+	// Synthesize our raftpb.ConfState from desc.
 	var cs raftpb.ConfState
-	cs.Nodes = []uint64{uint64(r.rm.RaftNodeID())}
-	// TODO(bdarnell): cs.Nodes must give the replicas as of the log index we
-	// put in the SnapshotMetadata. When we produce a snapshot of the current
-	// time, replace the previous line with the following block.
-	/*for _, rep := range r.Desc().Replicas {
+	for _, rep := range desc.Replicas {
```

Comment:
We specifically do not use r.Desc() here because we are not holding the metaLock and it might change out from under us while we build the snapshot. We must use the version of the descriptor in our rocksdb snapshot. (we could do a separate MVCCGet instead of reading it while iterating).